### PR TITLE
Add support for WebApplicationFactory

### LINKED ISSUE #435: Verify entityframework library with webapplicationfactory
#### Is the feature request related to a problem

It seems the verify entityframework library is not usable in tests with a webapplicationfactory. We write integration tests against a webapplicationfactory (https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.testing.webapplicationfactory-1?view=aspnetcore-5.0) to test our complete backend. During setup of the factory, we register our entityframework dbcontext as a singleton, so that we have the same instance outside of the test as inside the test call.
It seems that the verify entityframework library uses asynclocal to record all actions on the dbcontext. This has the final result that the recorded actions inside the api call are not available (and thus verifiable) in the integration test.

#### Describe the solution

Possibly an option / feature flag that the changes on the dbcontext are not recorded on an asynclocal variable, but just a normal (static) variable. This will not work in all scenario's where the library is currently used, but the extra option would enable the library to be used in the webapplication factory scenario.

#### Describe alternatives considered

We have considered the verify sqlserver library, however the verification of individual sql statements is too low level for us and would not be maintanable / verifiable for us in the long term.

### PR DESCRIPTION
**Pull request by OskarKlintrot**
Fixes #435

~~No test cases for `WebApplicationFactory` yet, I open this PR just because it's much easier to show the diff in a PR.~~

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/VerifyTests/Verify.EntityFramework?pullRequest=465) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/VerifyTests/Verify.EntityFramework?pullRequest=465) before we can accept your contribution.<br/><br/>:x: OskarKlintrot [sign now](https://cla.dotnetfoundation.org/VerifyTests/Verify.EntityFramework?pullRequest=465)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/VerifyTests/Verify.EntityFramework?pullRequest=465) it.</sub>

**Comment by OskarKlintrot** [comment]
@SimonCropp I accidently check the wrong radio button in the CLA, can you revoke my signing so I can redo it? I can't change it now unfortunately.

File: src/Verify.EntityFramework.Tests/CoreTests.cs [hunk 1]
```diff
@@ -327,6 +333,55 @@ await data.Companies
...
+
+                    services.AddScoped(_ => data);
+                });
+
+        protected override IHostBuilder CreateHostBuilder() =>
+            Host.CreateDefaultBuilder()
+                .ConfigureWebHostDefaults(webBuilder => webBuilder.UseStartup<Startup>());
+    }
+
+    public class Startup
```
Comment (OskarKlintrot) [inline src/Verify.EntityFramework.Tests/CoreTests.cs:369]: I tried to use minimal api but I didn't manage to solve it unless I created a new project, which seemed overkill. Unfortunately `WebApplicationFactory<T>` will be looking for the assembly entry point and there already is a (machine generated) entry point in test projects.

**Comment by SimonCropp** [comment]
if u run two tests in parallel, wont the results overlap?

**Comment by OskarKlintrot** [comment]
You're right, they will mess with each others data :(

**Comment by OskarKlintrot** [comment]
Now the tests can be run in parallel. It tok a lot of code and some new public facing api's though...

**Comment by OskarKlintrot** [comment]
I totally forgot that the interceptor is singleton in each WebApplicationFactory, they aren't sharing the same singleton instance. Now the PR looks a bit cleaner.

**Comment by OskarKlintrot** [comment]
I haven't written any docs btw, I will do it if you think this PR makes sense :)

**Comment by SimonCropp** [comment]
i think this looks good. if you add some docs we can merge and deploy

**Comment by OskarKlintrot** [comment]
Great! I added a section about `WebApplicationFactory`, I hope it makes sense! I changed a few variables in the tests to make the docs coherent.

**Pull request merged**